DS-4107 Metadata as map updates for Angular

cwilper opened the pull request:
Updates the angular code to consume and work with metadata modeled as a map. Some code consolidation was done here, so metadata-processing code is done in a common place, which is now well-covered by an included set of tests.

To work, this requires REST PR: https://github.com/DSpace/DSpace/pull/2287 which returns DSO metadata in the expected map form.

For more information, see https://jira.duraspace.org/browse/DS-4107

cwilper commented:
Rebased on latest master. This is ready for review and when approved, should be merged at the same time as the corresponding REST implementation: https://github.com/DSpace/DSpace/pull/2287

atarix83 commented on src/app/core/shared/metadata.model.ts:
```diff
@@ -0,0 +1,149 @@
+  /** Checks the given map for a matching value. */
+  public static has(mdMapOrMaps: MetadataMap | MetadataMap[], keyOrKeys: string | string[],
+                    filter?: MetadataValueFilter): boolean {
+    return Metadata.first(mdMapOrMaps, keyOrKeys, filter) !== undefined;
```
you could use isNotUndefined util method

atarix83 commented on src/app/shared/object-grid/search-result-grid-element/search-result-grid-element.component.ts:
```diff
@@ -22,39 +21,14 @@ export class SearchResultGridElementComponent<T extends SearchResult<K>, K exten
-      );
-    }
-    return results;
+  /** Gets all matching metadata string values from hitHighlights or dso metadata, preferring hitHighlights. */
```
Please use Supported TypeDoc tags to comment this method

atarix83 commented on src/app/shared/object-grid/search-result-grid-element/search-result-grid-element.component.ts:
```diff
@@ -22,39 +21,14 @@ export class SearchResultGridElementComponent<T extends SearchResult<K>, K exten
-      result = this.dso.findMetadata(key);
-    }
-    return result;
+  /** Gets the first matching metadata string value from hitHighlights or dso metadata, preferring hitHighlights. */
```
Please use Supported TypeDoc tags to comment this method

atarix83 commented on src/app/shared/object-list/search-result-list-element/search-result-list-element.component.ts:
```diff
@@ -22,39 +21,14 @@ export class SearchResultListElementComponent<T extends SearchResult<K>, K exten
-      );
-    }
-    return results;
+  /** Gets all matching metadata string values from hitHighlights or dso metadata, preferring hitHighlights. */
```
Please use Supported TypeDoc tags to comment this method

atarix83 commented on src/app/shared/object-list/search-result-list-element/search-result-list-element.component.ts:
```diff
@@ -22,39 +21,14 @@ export class SearchResultListElementComponent<T extends SearchResult<K>, K exten
-      result = this.dso.findMetadata(key);
-    }
-    return result;
+  /** Gets the first matching metadata string value from hitHighlights or dso metadata, preferring hitHighlights. */
```
Please use Supported TypeDoc tags to comment this method

cwilper commented on src/app/core/browse/browse.service.ts:
```diff
@@ -42,16 +42,16 @@ import { DSpaceObject } from '../shared/dspace-object.model';
 export class BrowseService {
```
Ok, I went ahead and added a basic class-level TypeDoc here, but notice that the change to this particular file was not a functional one...it was just renaming a variable for consistency with other touched code.

cwilper commented:
Thanks for the the review @atarix83 . I have rebased, resolving conflicts, and improved TypeDocs and used isUndefined/isNotUndefined as suggested.

cwilper commented:
Looks like checks have passed, so once @atarix83 approves, I believe this can be merged.

**Note: whoever merges this, please merge the REST impl side at the same time (already approved): https://github.com/DSpace/DSpace/pull/2287**

atarix83 commented:
@cwilper thanks for the changes.
It looks good to me now

paulo-graca commented on src/app/+item-page/edit-item-page/modify-item-overview/modify-item-overview.component.html:
```diff
@@ -7,10 +7,12 @@
+  <ng-container *ngFor="let entry of metadata | keyvalue">
+    <tr *ngFor="let value of entry.value" class="metadata-row">
+      <td>{{entry.key}}</td>
+      <td>{{value.value}}</td>
```
I dislike having `value.value` I suggest to rename `entry `to `mdItem` and `value` to `mdValue`. Then you will have `mdValue.value`

paulo-graca commented on src/app/+item-page/full/full-item-page.component.html:
```diff
@@ -9,11 +9,13 @@
-          <td>{{metadatum.value}}</td>
-          <td>{{metadatum.language}}</td>
-        </tr>
+          <ng-container *ngFor="let entry of (metadata$ | async) | keyvalue">
```
as I suggested, replacement of `entry` with `mdItem` and `value` with `mdValue`

paulo-graca commented:
This PR doesn't work properly, metadata isn't showed. I'm using atmire's REST endpoints. Please check my print-screen (https://drive.google.com/open?id=1bJypXyDXECJP7BlJ3LXsTGtR45dX2PXz).

cwilper commented:
@paulo-graca thanks for taking a look. I have adjusted variable naming to hopefully be more clear. I liked your suggestion about using `mdValue` and `mdValues` for clarity, and implemented that. You also suggested renaming `entry` to `mdItem`. I agree with making it more specific (prefix with md, consistent with the others), but I think "Entry" is really the right term here rather than "Item" -- it is a single _entry_ of a map, which contains a `key` and `value` property. So I called it `mdEntry`. Please let me know if that works ok for you.

> This PR doesn't work properly, metadata isn't showed.

This PR requires that you run with a specific REST PR or it won't work (see description). That PR is not yet running on Atmire's server and probably won't (permanently) until *both* PRs are merged. So I would suggest running a local copy of the the REST PR in order to test functionality.

paulo-graca commented:
👍 @cwilper thank you for considering my suggestions! I will try to do a new test with the dependency DSpace/DSpace#2287

tdonohue commented:
Already at +2, and it looks great to me too (did a quick scan of the code).  Merging.  Thanks @cwilper!
